In [1]:
import numpy as np
import pandas as pd


In [2]:
anime_data=pd.read_csv('../Data/anime.csv')

In [3]:
anime_data.tail()

,MAL_ID,Name,Score,Genres,English name,Japanese name,Type,Episodes,Aired,Premiered,...,Score-10,Score-9,Score-8,Score-7,Score-6,Score-5,Score-4,Score-3,Score-2,Score-1
17557,48481,Daomu Biji Zhi Qinling Shen Shu,Unknown,"Adventure, Mystery, Supernatural",Unknown,盗墓笔记之秦岭神树,ONA,Unknown,"Apr 4, 2021 to ?",Unknown,...,Unknown,Unknown,Unknown,1.0,Unknown,Unknown,Unknown,Unknown,Unknown,Unknown
17558,48483,Mieruko-chan,Unknown,"Comedy, Horror, Supernatural",Unknown,見える子ちゃん,TV,Unknown,2021 to ?,Unknown,...,Unknown,Unknown,Unknown,Unknown,Unknown,Unknown,Unknown,Unknown,Unknown,Unknown
17559,48488,Higurashi no Naku Koro ni Sotsu,Unknown,"Mystery, Dementia, Horror, Psychological, Supe...",Higurashi:When They Cry – SOTSU,ひぐらしのなく頃に卒,TV,Unknown,"Jul, 2021 to ?",Summer 2021,...,1.0,Unknown,Unknown,Unknown,Unknown,Unknown,Unknown,Unknown,Unknown,Unknown
17560,48491,Yama no Susume: Next Summit,Unknown,"Adventure, Slice of Life, Comedy",Unknown,ヤマノススメ Next Summit,TV,Unknown,Unknown,Unknown,...,Unknown,Unknown,Unknown,Unknown,Unknown,Unknown,Unknown,Unknown,Unknown,Unknown
17561,48492,Scarlet Nexus,Unknown,"Action, Fantasy",Unknown,SCARLET NEXUS,TV,Unknown,"Jul, 2021 to ?",Summer 2021,...,Unknown,Unknown,Unknown,Unknown,Unknown,Unknown,Unknown,Unknown,Unknown,Unknown


In [4]:
anime_data.shape

(17562, 35)

In [31]:
anime_data.info()

<class 'pandas.DataFrame'>
RangeIndex: 17562 entries, 0 to 17561
Data columns (total 35 columns):
 #   Column         Non-Null Count  Dtype
---  ------         --------------  -----
 0   MAL_ID         17562 non-null  int64
 1   Name           17562 non-null  str  
 2   Score          17562 non-null  str  
 3   Genres         17562 non-null  str  
 4   English name   17562 non-null  str  
 5   Japanese name  17562 non-null  str  
 6   Type           17562 non-null  str  
 7   Episodes       17562 non-null  str  
 8   Aired          17562 non-null  str  
 9   Premiered      17562 non-null  str  
 10  Producers      17562 non-null  str  
 11  Licensors      17562 non-null  str  
 12  Studios        17562 non-null  str  
 13  Source         17562 non-null  str  
 14  Duration       17562 non-null  str  
 15  Rating         17562 non-null  str  
 16  Ranked         17562 non-null  str  
 17  Popularity     17562 non-null  int64
 18  Members        17562 non-null  int64
 19  Favorites      

In [35]:
anime_data.dtypes

MAL_ID           int64
Name               str
Score              str
Genres             str
English name       str
Japanese name      str
Type               str
Episodes           str
Aired              str
Premiered          str
Producers          str
Licensors          str
Studios            str
Source             str
Duration           str
Rating             str
Ranked             str
Popularity       int64
Members          int64
Favorites        int64
Watching         int64
Completed        int64
On-Hold          int64
Dropped          int64
Plan to Watch    int64
Score-10           str
Score-9            str
Score-8            str
Score-7            str
Score-6            str
Score-5            str
Score-4            str
Score-3            str
Score-2            str
Score-1            str
dtype: object

In [6]:
anime_data.describe()

,MAL_ID,Popularity,Members,Favorites,Watching,Completed,On-Hold,Dropped,Plan to Watch
count,17562.000000,17562.000000,1.756200e+04,17562.000000,17562.000000,1.756200e+04,17562.000000,17562.000000,17562.000000
mean,21477.192347,8763.452340,3.465854e+04,457.746270,2231.487758,2.209557e+04,955.049653,1176.599533,8199.831227
std,14900.093170,5059.327278,1.252821e+05,4063.473313,14046.688133,9.100919e+04,4275.675096,4740.348653,23777.691963
min,1.000000,0.000000,1.000000e+00,0.000000,0.000000,0.000000e+00,0.000000,0.000000,1.000000
25%,5953.500000,4383.500000,3.360000e+02,0.000000,13.000000,1.110000e+02,6.000000,37.000000,112.000000
50%,22820.000000,8762.500000,2.065000e+03,3.000000,73.000000,8.175000e+02,45.000000,77.000000,752.500000
75%,35624.750000,13145.000000,1.322325e+04,31.000000,522.000000,6.478000e+03,291.750000,271.000000,4135.500000
max,48492.000000,17565.000000,2.589552e+06,183914.000000,887333.000000,2.182587e+06,187919.000000,174710.000000,425531.000000


In [7]:
anime_data.columns

Index(['MAL_ID', 'Name', 'Score', 'Genres', 'English name', 'Japanese name',
       'Type', 'Episodes', 'Aired', 'Premiered', 'Producers', 'Licensors',
       'Studios', 'Source', 'Duration', 'Rating', 'Ranked', 'Popularity',
       'Members', 'Favorites', 'Watching', 'Completed', 'On-Hold', 'Dropped',
       'Plan to Watch', 'Score-10', 'Score-9', 'Score-8', 'Score-7', 'Score-6',
       'Score-5', 'Score-4', 'Score-3', 'Score-2', 'Score-1'],
      dtype='str')

In [36]:
print(anime_data[['Score', 'Ranked',
                  'Score-1', 'Score-2', 'Score-3', 'Score-4',
                  'Score-5', 'Score-6', 'Score-7', 'Score-8',
                  'Score-9', 'Score-10']].dtypes)

Score       str
Ranked      str
Score-1     str
Score-2     str
Score-3     str
Score-4     str
Score-5     str
Score-6     str
Score-7     str
Score-8     str
Score-9     str
Score-10    str
dtype: object


In [8]:
anime_data.isnull().sum()

MAL_ID           0
Name             0
Score            0
Genres           0
English name     0
Japanese name    0
Type             0
Episodes         0
Aired            0
Premiered        0
Producers        0
Licensors        0
Studios          0
Source           0
Duration         0
Rating           0
Ranked           0
Popularity       0
Members          0
Favorites        0
Watching         0
Completed        0
On-Hold          0
Dropped          0
Plan to Watch    0
Score-10         0
Score-9          0
Score-8          0
Score-7          0
Score-6          0
Score-5          0
Score-4          0
Score-3          0
Score-2          0
Score-1          0
dtype: int64

In [9]:
anime_data.duplicated().sum()

np.int64(0)

In [34]:
filtering_unknown=(anime_data=="Unknown").sum()
filtering_unknown

MAL_ID               0
Name                 0
Score             5141
Genres              63
English name     10565
Japanese name       48
Type                37
Episodes           516
Aired              309
Premiered        12817
Producers         7794
Licensors        13616
Studios           7079
Source            3567
Duration           555
Rating             688
Ranked            1762
Popularity           0
Members              0
Favorites            0
Watching             0
Completed            0
On-Hold              0
Dropped              0
Plan to Watch        0
Score-10           437
Score-9           3167
Score-8           1371
Score-7            503
Score-6            511
Score-5            584
Score-4            977
Score-3           1307
Score-2           1597
Score-1            459
dtype: int64

In [29]:
anime_data.head().T

,0,1,2,3,4
MAL_ID,1,5,6,7,8
Name,Cowboy Bebop,Cowboy Bebop: Tengoku no Tobira,Trigun,Witch Hunter Robin,Bouken Ou Beet
Score,8.78,8.39,8.24,7.27,6.98
Genres,"Action, Adventure, Comedy, Drama, Sci-Fi, Space","Action, Drama, Mystery, Sci-Fi, Space","Action, Sci-Fi, Adventure, Comedy, Drama, Shounen","Action, Mystery, Police, Supernatural, Drama, ...","Adventure, Fantasy, Shounen, Supernatural"
English name,Cowboy Bebop,Cowboy Bebop:The Movie,Trigun,Witch Hunter Robin,Beet the Vandel Buster
Japanese name,カウボーイビバップ,カウボーイビバップ 天国の扉,トライガン,Witch Hunter ROBIN (ウイッチハンターロビン),冒険王ビィト
Type,TV,Movie,TV,TV,TV
Episodes,26,1,26,26,52
Aired,"Apr 3, 1998 to Apr 24, 1999","Sep 1, 2001","Apr 1, 1998 to Sep 30, 1998","Jul 2, 2002 to Dec 24, 2002","Sep 30, 2004 to Sep 29, 2005"
Premiered,Spring 1998,Unknown,Spring 1998,Summer 2002,Fall 2004


| Data-quality check          | Finding |
| --------------------------- | ------- |
| Number of rows              | 17562       |
| Number of columns           | 35    |
| Duplicate rows              | 0       |
| Columns with missing values | 0       |
| Number of Columns containig `Unknown`| 25      |


### Data Quality Assessment

The dataset contains 17,562 rows and 35 columns. No duplicate rows were identified, and Pandas reported no explicit `NaN` values.

However, further investigation revealed that 25 columns contain `"Unknown"` values. The number of unknown entries varies substantially across columns, with `Premiered`, `Licensors`, and `English name` containing particularly large numbers of unknown values.

The presence of `"Unknown"` does not necessarily indicate invalid data. Each affected column will therefore be evaluated individually to determine whether the value should be retained as a category, converted to `NaN`, or handled through another appropriate cleaning strategy.

#### Before any convertion, we want to ensure that the columns to be converted has not other error apart from the Unknown

In [47]:
anime_data['Score'].unique()

<StringArray>
['8.78', '8.39', '8.24', '7.27', '6.98', '7.95', '8.06', '7.59', '8.15',
 '8.76',
 ...
 '8.86', '8.84', '8.71', '8.62', '8.68', '2.18', '8.88', '9.17', '3.89',
 '4.06']
Length: 533, dtype: str

In [ ]:
anime_data['Score'].unique()[:20]

<StringArray>
[  '28.0',  '159.0',  '266.0', '2481.0', '3710.0',  '604.0',  '468.0',
 '1317.0',  '360.0',   '30.0',  '660.0',   '95.0',  '675.0', '6493.0',
  '625.0', '1844.0',  '896.0', '2240.0',  '609.0', '1565.0']
Length: 20, dtype: str

In [49]:
anime_data['Score'].unique()[-20:]

<StringArray>
['8.87', '4.34', '8.79', '3.84', '3.42', '4.22', '3.58', '8.99', '8.93',
 '8.81', '8.86', '8.84', '8.71', '8.62', '8.68', '2.18', '8.88', '9.17',
 '3.89', '4.06']
Length: 20, dtype: str

In [51]:
anime_data["Score"].value_counts().head(20)

Score
Unknown    5141
6.48         74
6.31         72
6.3          72
6.52         71
6.45         70
6.65         68
6.19         65
6.6          65
6.07         65
6.23         64
6.54         64
6.59         63
6.11         63
6.03         63
6.33         63
6.55         62
6.82         62
6.12         62
6.13         62
Name: count, dtype: int64

In [52]:
anime_data["Ranked"].unique()

<StringArray>
[   '28.0',   '159.0',   '266.0',  '2481.0',  '3710.0',   '604.0',   '468.0',
  '1317.0',   '360.0',    '30.0',
 ...
 '13197.0', '15477.0', '13687.0',  '6360.0',  '5461.0', '15206.0', '14658.0',
 '10760.0', '13254.0', '13980.0']
Length: 10490, dtype: str

In [56]:
anime_data['Ranked'].value_counts()

Ranked
Unknown    1762
5297.0        6
12978.0       6
1804.0        5
5946.0        5
           ... 
15206.0       1
14658.0       1
10760.0       1
13254.0       1
13980.0       1
Name: count, Length: 10490, dtype: int64

In [57]:
anime_data["Score-10"].head()

0    229170.0
1     30043.0
2     50229.0
3      2182.0
4       312.0
Name: Score-10, dtype: str

In [59]:
anime_data['Score-10'].unique()[20:50]

<StringArray>
['186797.0',  '13516.0', '127081.0',  '58627.0',  '56237.0',  '39845.0',
  '38050.0',   '2958.0',  '77192.0',   '5266.0',   '2021.0',   '5425.0',
   '1464.0',   '1644.0',   '3124.0',   '2107.0',    '174.0',    '166.0',
  '23840.0',   '1451.0',  '20499.0',  '11469.0',   '7515.0',   '1542.0',
   '3464.0',   '7868.0',   '5500.0',  '18422.0',   '6677.0',   '8654.0']
Length: 30, dtype: str

In [60]:
anime_data['Score-10'].value_counts()

Score-10
4.0       936
5.0       902
3.0       751
6.0       750
7.0       616
         ... 
2926.0      1
7564.0      1
795.0       1
4243.0      1
1305.0      1
Name: count, Length: 3379, dtype: int64

In [64]:
anime_data['Score-10'].dtypes

<StringDtype(storage='python', na_value=nan)>

#### Testing whether unknown is the only missing value before conversion

In [ ]:
pd.to_numeric(anime_data['Score-10'], errors='coerce').isna().sum()

np.int64(437)

In [66]:
(anime_data['Score-10'] == 'Unknown').sum()

np.int64(437)

#### Test whether `Unknown` is the only thing preventing the columns, Ranked and Score from being numerical

In [ ]:
for column in ['Score',"Ranked"]:
    unknown=(anime_data[column]).sum()
    non_numerical=pd.to_numeric(anime_data[column], errors='coerce').isna().sum()

    print(column)
    print("Unknown: ",unknown)
    print("Non-numeric: ", non_numerical)